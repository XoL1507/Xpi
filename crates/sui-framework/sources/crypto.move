// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

/// Library for cryptography onchain.
module sui::crypto {
    friend sui::validator;

    /// @param signature: A 65-bytes signature in form (r, s, v) that is signed using 
    /// Secp256k1. Reference implementation on signature generation using RFC6979: 
    /// https://github.com/MystenLabs/narwhal/blob/5d6f6df8ccee94446ff88786c0dbbc98be7cfc09/crypto/src/secp256k1.rs
    /// The accepted v values are {0, 1, 2, 3}.
    /// 
    /// @param hashed_msg: the hashed 32-bytes message. The message must be hashed instead 
    /// of plain text to be secure.
    /// 
    /// If the signature is valid, return the corresponding recovered Secpk256k1 public 
    /// key, otherwise throw error. This is similar to ecrecover in Ethereum, can only be 
    /// applied to Secp256k1 signatures.
    public native fun ecrecover(signature: vector<u8>, hashed_msg: vector<u8>): vector<u8>;

    /// @param data: arbitrary bytes data to hash
    /// Hash the input bytes using keccak256 and returns 32 bytes.
    public native fun keccak256(data: vector<u8>): vector<u8>;

    /// @param signature: A 48-bytes signature that is a point on the G1 subgroup
    /// @param public_key: A 96-bytes public key that is a point on the G2 subgroup
    /// @param msg: The message that we test the signature against.
    ///
    /// If the signature is a valid BLS12381 signature of the message and public key, return true.
    /// Otherwise, return false.
    public native fun bls12381_verify_g1_sig(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>): bool; 

    /// @param signature: A 65-bytes signature in form (r, s, v) that is signed using 
    /// Secp256k1. Reference implementation on signature generation using RFC6979: 
    /// https://github.com/MystenLabs/narwhal/blob/5d6f6df8ccee94446ff88786c0dbbc98be7cfc09/crypto/src/secp256k1.rs
    /// 
    /// @param public_key: The public key to verify the signature against
    /// @param hashed_msg: The hashed 32-bytes message, same as what the signature is signed against.
    /// 
    /// If the signature is valid to the pubkey and hashed message, return true. Else false.
    public native fun secp256k1_verify(signature: vector<u8>, public_key: vector<u8>, hashed_msg: vector<u8>): bool;

    use sui::elliptic_curve::{Self as ec, RistrettoPoint};

    /// Only bit_length = 64, 32, 16, 8 will work.
    native fun native_verify_full_range_proof(proof: vector<u8>, commitment: vector<u8>, bit_length: u64);

    /// @param proof: The bulletproof
    /// @param commitment: The commitment which we are trying to verify the range proof for
    /// @param bit_length: The bit length that we prove the committed value is whithin. Note that bit_length must be either 64, 32, 16, or 8.
    /// 
    /// If the range proof is valid, execution succeeds, else panics.
    public fun verify_full_range_proof(proof: vector<u8>, commitment: RistrettoPoint, bit_length: u64) {
        native_verify_full_range_proof(proof, ec::bytes(&commitment), bit_length)
    }

    /// @param signature: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param public_key: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param msg: The message that we test the signature against.
    /// 
    /// If the signature is a valid Ed25519 signature of the message and public key, return true.
    /// Otherwise, return false.
    public(friend) native fun ed25519_verify(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>): bool;

    /// @param signature: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param public_key: 32-byte signature that is a point on the Ed25519 elliptic curve.
    /// @param msg: The message that we test the signature against.
    /// @param domain: The domain that the signature is tested again. We essentially prepend this to the message.
    /// 
    /// If the signature is a valid Ed25519 signature of the message and public key, return true.
    /// Otherwise, return false.
    public(friend) fun ed25519_verify_with_domain(signature: vector<u8>, public_key: vector<u8>, msg: vector<u8>, domain: vector<u8>): bool {
        std::vector::append(&mut domain, msg);
        ed25519_verify(signature, public_key, domain)
    }
}
