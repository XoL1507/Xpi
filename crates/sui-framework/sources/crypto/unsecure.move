// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

module sui::unsecure {
    friend sui::validator;

    public native fun unsecure_verify(signature: &vector<u8>, public_key: &vector<u8>, msg: &vector<u8>): bool;

    public(friend) fun unsecure_verify_with_domain(
        signature: &vector<u8>,
        public_key: &vector<u8>,
        msg: vector<u8>,
        domain: vector<u8>
    ): bool {
        std::vector::append(&mut domain, msg);
        unsecure_verify(signature, public_key, &domain)
    }

}
