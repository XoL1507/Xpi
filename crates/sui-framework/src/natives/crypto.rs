// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0
use crate::{legacy_emit_cost, legacy_empty_cost};
use curve25519_dalek_ng::scalar::Scalar;
use fastcrypto::{
    bls12381::{min_pk, min_sig},
    bulletproofs::{BulletproofsRangeProof, PedersenCommitment},
    ed25519::{Ed25519PublicKey, Ed25519Signature},
    hmac,
    secp256k1::{Secp256k1PublicKey, Secp256k1Signature},
    traits::ToFromBytes,
    Verifier,
};
use fastcrypto_zkp::{verifier::{PreparedVerifyingKey, verify_with_processed_vk, Proof}, conversions::{PrimeField, bls_g1_affine_from_zcash_bytes, bls_g2_affine_from_zcash_bytes, BlsFr, BlsG1Affine}};
use move_binary_format::errors::PartialVMResult;
use move_vm_runtime::native_functions::NativeContext;
use move_vm_types::{
    loaded_data::runtime_types::Type,
    natives::function::NativeResult,
    pop_arg,
    values::{Value, VectorRef},
};
use smallvec::smallvec;
use std::collections::VecDeque;
use sui_types::error::SuiError;

pub const FAIL_TO_RECOVER_PUBKEY: u64 = 0;
pub const INVALID_SIGNATURE: u64 = 1;
pub const INVALID_BULLETPROOF: u64 = 2;
pub const INVALID_RISTRETTO_GROUP_ELEMENT: u64 = 3;
pub const INVALID_RISTRETTO_SCALAR: u64 = 4;
pub const BULLETPROOFS_VERIFICATION_FAILED: u64 = 5;
pub const INVALID_PUBKEY: u64 = 6;

/// Using the word "sui" for nothing-up-my-sleeve number guarantees.
pub const BP_DOMAIN: &[u8] = b"sui";

/// Native implementation of ecrecover in public Mo
/// ve API, see crypto.move for specifications.
pub fn ecrecover(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 2);

    let hashed_msg = pop_arg!(args, VectorRef);
    let signature = pop_arg!(args, VectorRef);

    let hashed_msg_ref = hashed_msg.as_bytes_ref();
    let signature_ref = signature.as_bytes_ref();

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
    let cost = legacy_empty_cost();
    match recover_pubkey(&signature_ref, &hashed_msg_ref) {
        Ok(pubkey) => Ok(NativeResult::ok(
            cost,
            smallvec![Value::vector_u8(pubkey.as_bytes().to_vec())],
        )),
        Err(SuiError::InvalidSignature { error: _ }) => {
            Ok(NativeResult::err(cost, INVALID_SIGNATURE))
        }
        Err(_) => Ok(NativeResult::err(cost, FAIL_TO_RECOVER_PUBKEY)),
    }
}

fn recover_pubkey(signature: &[u8], hashed_msg: &[u8]) -> Result<Secp256k1PublicKey, SuiError> {
    match <Secp256k1Signature as ToFromBytes>::from_bytes(signature) {
        Ok(signature) => match signature.recover(hashed_msg) {
            Ok(pubkey) => Ok(pubkey),
            Err(e) => Err(SuiError::KeyConversionError(e.to_string())),
        },
        Err(e) => Err(SuiError::InvalidSignature {
            error: e.to_string(),
        }),
    }
}

/// Convert a compressed 33-bytes Secp256k1 pubkey to an 65-bytes uncompressed one.
pub fn decompress_pubkey(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 1);

    let pubkey = pop_arg!(args, VectorRef);
    let pubkey_ref = pubkey.as_bytes_ref();

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
    let cost = legacy_empty_cost();

    match Secp256k1PublicKey::from_bytes(&pubkey_ref) {
        Ok(pubkey) => {
            let uncompressed = &pubkey.pubkey.serialize_uncompressed();
            Ok(NativeResult::ok(
                cost,
                smallvec![Value::vector_u8(uncompressed.to_vec())],
            ))
        }
        Err(_) => Ok(NativeResult::err(cost, INVALID_PUBKEY)),
    }
}

/// Native implemention of keccak256 in public Move API, see crypto.move for specifications.
pub fn keccak256(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 1);

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
    let cost = legacy_empty_cost();
    let msg = pop_arg!(args, VectorRef);
    let msg_ref = msg.as_bytes_ref();

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(
            <sha3::Keccak256 as sha3::digest::Digest>::digest(&*msg_ref)
                .as_slice()
                .to_vec()
        )],
    ))
}

/// Native implemention of secp256k1_verify in public Move API, see crypto.move for specifications.
pub fn secp256k1_verify(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 3);

    let hashed_msg = pop_arg!(args, VectorRef);
    let public_key_bytes = pop_arg!(args, VectorRef);
    let signature_bytes = pop_arg!(args, VectorRef);

    let hashed_msg_ref = hashed_msg.as_bytes_ref();
    let public_key_bytes_ref = public_key_bytes.as_bytes_ref();
    let signature_bytes_ref = signature_bytes.as_bytes_ref();

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/4086
    let cost = legacy_emit_cost();

    let signature = match <Secp256k1Signature as ToFromBytes>::from_bytes(&signature_bytes_ref) {
        Ok(signature) => signature,
        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    };

    let public_key = match <Secp256k1PublicKey as ToFromBytes>::from_bytes(&public_key_bytes_ref) {
        Ok(public_key) => public_key,
        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    };

    match public_key.verify_hashed(&hashed_msg_ref, &signature) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(true)])),
        Err(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    }
}

/// Native implementation of bls12381_min_sig_verify in public Move API, see bls12381.move for specifications.
/// Note that this function only works for signatures in G1 and public keys in G2.
pub fn bls12381_min_sig_verify(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 3);

    let msg = pop_arg!(args, VectorRef);
    let public_key_bytes = pop_arg!(args, VectorRef);
    let signature_bytes = pop_arg!(args, VectorRef);

    let msg_ref = msg.as_bytes_ref();
    let public_key_bytes_ref = public_key_bytes.as_bytes_ref();
    let signature_bytes_ref = signature_bytes.as_bytes_ref();

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3868
    let cost = legacy_emit_cost();

    let signature =
        match <min_sig::BLS12381Signature as ToFromBytes>::from_bytes(&signature_bytes_ref) {
            Ok(signature) => signature,
            Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
        };

    let public_key =
        match <min_sig::BLS12381PublicKey as ToFromBytes>::from_bytes(&public_key_bytes_ref) {
            Ok(public_key) => public_key,
            Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
        };

    match public_key.verify(&msg_ref, &signature) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(true)])),
        Err(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    }
}

/// Native implementation of bls12381_min_pk_verify in public Move API, see bls12381.move for specifications.
/// Note that this function only works for signatures in G2 and public keys in G1.
pub fn bls12381_min_pk_verify(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 3);

    let msg = pop_arg!(args, VectorRef);
    let public_key_bytes = pop_arg!(args, VectorRef);
    let signature_bytes = pop_arg!(args, VectorRef);

    let msg_ref = msg.as_bytes_ref();
    let public_key_bytes_ref = public_key_bytes.as_bytes_ref();
    let signature_bytes_ref = signature_bytes.as_bytes_ref();

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3868
    let cost = legacy_emit_cost();

    let signature =
        match <min_pk::BLS12381Signature as ToFromBytes>::from_bytes(&signature_bytes_ref) {
            Ok(signature) => signature,
            Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
        };

    let public_key =
        match <min_pk::BLS12381PublicKey as ToFromBytes>::from_bytes(&public_key_bytes_ref) {
            Ok(public_key) => public_key,
            Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
        };

    match public_key.verify(&msg_ref, &signature) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(true)])),
        Err(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    }
}

/// Native implementation of Bulletproofs range proof in public Move API, see crypto.move for
/// specifications.
pub fn verify_range_proof(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 3);

    let bit_length = pop_arg!(args, u64);
    let commitment_bytes = pop_arg!(args, VectorRef);
    let proof_bytes = pop_arg!(args, VectorRef);
    let cost = legacy_empty_cost();

    let commitment_bytes_ref = commitment_bytes.as_bytes_ref();
    let proof_bytes_ref = proof_bytes.as_bytes_ref();

    let proof = if let Ok(val) = BulletproofsRangeProof::from_bytes(&proof_bytes_ref) {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_BULLETPROOF));
    };

    let commitment = if let Ok(val) = PedersenCommitment::from_bytes(&commitment_bytes_ref) {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_GROUP_ELEMENT));
    };

    match proof.verify_bit_length(&commitment, bit_length as usize, BP_DOMAIN) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![])),
        _ => Ok(NativeResult::err(cost, BULLETPROOFS_VERIFICATION_FAILED)),
    }
}

pub fn add_ristretto_point(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 2);

    let point_a = pop_arg!(args, Vec<u8>);
    let point_b = pop_arg!(args, Vec<u8>);
    let cost = legacy_empty_cost();

    let rist_point_a = if let Ok(val) = PedersenCommitment::from_bytes(&point_a[..]) {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_GROUP_ELEMENT));
    };
    let rist_point_b = if let Ok(val) = PedersenCommitment::from_bytes(&point_b[..]) {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_GROUP_ELEMENT));
    };

    let sum = rist_point_a + rist_point_b;

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(sum.as_bytes().to_vec())],
    ))
}

pub fn subtract_ristretto_point(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 2);

    let point_b = pop_arg!(args, Vec<u8>);
    let point_a = pop_arg!(args, Vec<u8>);
    let cost = legacy_empty_cost();

    let rist_point_a = if let Ok(val) = PedersenCommitment::from_bytes(&point_a[..]) {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_GROUP_ELEMENT));
    };
    let rist_point_b = if let Ok(val) = PedersenCommitment::from_bytes(&point_b[..]) {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_GROUP_ELEMENT));
    };

    let sum = rist_point_a - rist_point_b;

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(sum.as_bytes().to_vec())],
    ))
}

pub fn pedersen_commit(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 2);

    let blinding_factor_vec = pop_arg!(args, Vec<u8>);
    let value_vec = pop_arg!(args, Vec<u8>);
    let cost = legacy_empty_cost();

    let blinding_factor: [u8; 32] = if let Ok(val) = blinding_factor_vec.try_into() {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_SCALAR));
    };

    let value: [u8; 32] = if let Ok(val) = value_vec.try_into() {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_SCALAR));
    };

    let commitment = PedersenCommitment::new(value, blinding_factor);

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(commitment.as_bytes().to_vec())],
    ))
}

pub fn scalar_from_u64(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 1);

    let value = pop_arg!(args, u64);
    let cost = legacy_empty_cost();
    let scalar = Scalar::from(value);

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(scalar.as_bytes().to_vec())],
    ))
}

pub fn scalar_from_bytes(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 1);

    let value = pop_arg!(args, Vec<u8>);
    let cost = legacy_empty_cost();

    let value: [u8; 32] = if let Ok(val) = value.try_into() {
        val
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_SCALAR));
    };

    let scalar = if let Some(value) = Scalar::from_canonical_bytes(value) {
        value
    } else {
        return Ok(NativeResult::err(cost, INVALID_RISTRETTO_SCALAR));
    };

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(scalar.as_bytes().to_vec())],
    ))
}

/// Native implemention of ed25519_verify in public Move API, see crypto.move for specifications.
pub fn ed25519_verify(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 3);

    let msg = pop_arg!(args, VectorRef);
    let msg_ref = msg.as_bytes_ref();
    let public_key_bytes = pop_arg!(args, VectorRef);
    let public_key_bytes_ref = public_key_bytes.as_bytes_ref();
    let signature_bytes = pop_arg!(args, VectorRef);
    let signature_bytes_ref = signature_bytes.as_bytes_ref();

    let cost = legacy_empty_cost();

    let signature = match <Ed25519Signature as ToFromBytes>::from_bytes(&signature_bytes_ref) {
        Ok(signature) => signature,
        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    };

    let public_key = match <Ed25519PublicKey as ToFromBytes>::from_bytes(&public_key_bytes_ref) {
        Ok(public_key) => public_key,
        Err(_) => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    };

    match public_key.verify(&msg_ref, &signature) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(true)])),
        Err(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    }
}

/// Native implementation of hmac-sha3-256 in public Move API, see hmac.move for specifications.
pub fn hmac_sha3_256(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(ty_args.is_empty());
    debug_assert!(args.len() == 2);

    let message = pop_arg!(args, VectorRef);
    let key = pop_arg!(args, VectorRef);
    let hmac_key = hmac::HmacKey::from_bytes(&key.as_bytes_ref()).unwrap();

    // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
    let cost = legacy_empty_cost();

    Ok(NativeResult::ok(
        cost,
        smallvec![Value::vector_u8(
            hmac::hmac_sha3_256(&hmac_key, &message.as_bytes_ref()).to_vec()
        )],
    ))
}

/// Native implemention of ed25519_verify in public Move API, see crypto.move for specifications.
pub fn internal_verify_groth16_proof(
    _context: &mut NativeContext,
    ty_args: Vec<Type>,
    mut args: VecDeque<Value>,
) -> PartialVMResult<NativeResult> {
    debug_assert!(args.len() == 8);

    let vk_gamma_abc_g1_bytes = pop_arg!(args, Vec<u8>);
    let alpha_g1_beta_g2_bytes = pop_arg!(args, Vec<u8>);
    let gamma_g2_neg_pc_bytes = pop_arg!(args, Vec<u8>);
    let delta_g2_neg_pc_bytes = pop_arg!(args, Vec<u8>);
    
    let gamma_g2_neg_pc: [u8; 96] = gamma_g2_neg_pc_bytes.try_into().unwrap();
    let delta_g2_neg_pc: [u8; 96] = delta_g2_neg_pc_bytes.try_into().unwrap();

    let x_bytes = pop_arg!(args, Vec<u8>);
    let a_bytes = pop_arg!(args, Vec<u8>);
    let b_bytes = pop_arg!(args, Vec<u8>);
    let c_bytes = pop_arg!(args, Vec<u8>);

    let pvk = build_pvk(x_bytes.len(), vk_gamma_abc_g1_bytes, alpha_g1_beta_g2_bytes, &gamma_g2_neg_pc, &delta_g2_neg_pc).unwrap();

    let x_bytes = pop_arg!(args, Vec<u8>);
    let x: [u8; 48] = x_bytes.try_into().unwrap();

    let x = BlsFr::from_le_bytes_mod_order(&x);

    let a_bytes_arr: [u8; 48] = a_bytes.try_into().unwrap();
    let b_bytes_arr: [u8; 96] = b_bytes.try_into().unwrap();
    let c_bytes_arr: [u8; 48] = c_bytes.try_into().unwrap();

    let proof = Proof {
        a: bls_g1_affine_from_zcash_bytes(&a_bytes_arr).unwrap(),
        b: bls_g2_affine_from_zcash_bytes(&b_bytes_arr).unwrap(),
        c: bls_g1_affine_from_zcash_bytes(&c_bytes_arr).unwrap(),
    };

    let cost = legacy_empty_cost();

    match verify_with_processed_vk(&pvk, x, &proof) {
        Ok(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(true)])),
        Err(_) => Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
    }
}

fn build_pvk(input_size: usize, vk_gamma_abc_g1: Vec<u8>, alpha_g1_beta_g2_bytes: Vec<u8>, gamma_g2_neg_pc: &[u8; 96], delta_g2_neg_pc: &[u8; 96]) -> Result<PreparedVerifyingKey, SuiError> {
    let mut vec_points: Vec<BlsG1Affine> = Vec::new();
    if input_size * 48 != vk_gamma_abc_g1.len() {
        return Err(SuiError::InvalidPrivateKey);
    }
    (0..input_size.into()).step_by(48).for_each(
        |i| {
            let arr: [u8; 48] = vk_gamma_abc_g1[i..i+48].try_into().unwrap();
            vec_points.push(bls_g1_affine_from_zcash_bytes(&arr).unwrap());
        }
    );
    
    Ok(PreparedVerifyingKey {
        vk_gamma_abc_g1: vec_points,
        alpha_g1_beta_g2: alpha_g1_beta_g2_bytes.to_vec(),
        gamma_g2_neg_pc: bls_g2_affine_from_zcash_bytes(gamma_g2_neg_pc).unwrap(),
        delta_g2_neg_pc: bls_g2_affine_from_zcash_bytes(delta_g2_neg_pc).unwrap(),
    })
}