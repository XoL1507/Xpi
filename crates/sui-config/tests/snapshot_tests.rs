// Copyright (c) 2022, Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

// This file contains tests that detect changes in Sui configs.
// If a PR breaks one or more tests here, the PR probably has a real impact
// on a production configuration file. When test failure happens, the PR should
// be marked as a breaking change and reviewers should be aware of this.
//
// Owners and operators of production configuration files can add themselves to
// .github/CODEOWNERS for the corresponding snapshot tests, so they can get notified
// of changes. PRs that modifies snapshot files should wait for reviews from
// code owners (if any) before merging.
//
// To review snapshot changes, and fix snapshot differences,
// 0. Install cargo-insta
// 1. Run `cargo insta test --review` under `./sui-config`.
// 2. Review, accept or reject changes.

use std::net::{IpAddr, Ipv4Addr, SocketAddr};
use std::path::PathBuf;

use insta::assert_yaml_snapshot;
use multiaddr::Multiaddr;
use narwhal_crypto::traits::KeyPair;
use rand::rngs::StdRng;
use rand::SeedableRng;
use sui_config::{genesis::Builder, genesis_config::GenesisConfig};
use sui_config::{NetworkConfig, ValidatorInfo};
use sui_types::base_types::{ObjectID, SuiAddress};
use sui_types::crypto::{get_key_pair_from_rng, AuthorityKeyPair, AuthorityPublicKeyBytes};

#[test]
fn genesis_config_snapshot_matches() {
    // Test creating fake SuiAddress from PublicKeyBytes.
    let keypair: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let public_key = AuthorityPublicKeyBytes::from(keypair.public());
    let fake_addr = SuiAddress::from(&public_key);

    let fake_obj_id = ObjectID::from(fake_addr);
    let mut genesis_config = GenesisConfig::for_local_testing();
    for account in &mut genesis_config.accounts {
        account.address = Some(fake_addr);
        for gas_obj in &mut account.gas_objects {
            gas_obj.object_id = fake_obj_id;
        }
    }
    assert_yaml_snapshot!(genesis_config);
}

// If a Move change breaks the test below, it is because related Move functions are included
// in the genesis blob. The test failure can be fixed by updating the snapshot / blob.
// Also, folks in the file's CODEWOWNERS should be notified about the change. They may need to
// re-create the genesis blob.
#[test]
fn empty_genesis_snapshot_matches() {
    let genesis = Builder::new().build();
    assert_yaml_snapshot!(genesis);
}

#[test]
fn populated_genesis_snapshot_matches() {
    let genesis_config = GenesisConfig::for_local_testing();
    let (_account_keys, objects) = genesis_config
        .generate_accounts(&mut StdRng::from_seed([0; 32]))
        .unwrap();
    let key: AuthorityKeyPair = get_key_pair_from_rng(&mut StdRng::from_seed([0; 32])).1;
    let validator = ValidatorInfo {
        name: "0".into(),
        public_key: key.public().into(),
        stake: 1,
        delegation: 0,
        gas_price: 1,
        network_address: Multiaddr::empty(),
        narwhal_primary_to_primary: Multiaddr::empty(),
        narwhal_worker_to_primary: Multiaddr::empty(),
        narwhal_primary_to_worker: Multiaddr::empty(),
        narwhal_worker_to_worker: Multiaddr::empty(),
        narwhal_consensus_address: Multiaddr::empty(),
    };

    let genesis = Builder::new()
        .add_objects(objects)
        .add_validator(validator)
        .build();
    assert_yaml_snapshot!(genesis.validator_set());
    assert_yaml_snapshot!(genesis.committee().unwrap());
    assert_yaml_snapshot!(genesis.narwhal_committee());
    assert_yaml_snapshot!(genesis.sui_system_object());
    // Serialized `genesis` is not static and cannot be snapshot tested.
}

#[test]
fn network_config_snapshot_matches() {
    let temp_dir = tempfile::tempdir().unwrap();
    let committee_size = 7;
    let rng = StdRng::from_seed([0; 32]);
    let mut network_config = NetworkConfig::generate_with_rng(temp_dir.path(), committee_size, rng);
    // TODO: Inject static temp path and port numbers, instead of clearing them.
    for mut validator_config in &mut network_config.validator_configs {
        validator_config.db_path = PathBuf::from("/tmp/foo/");
        validator_config.network_address = Multiaddr::empty();
        let fake_socket = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)), 1);
        validator_config.json_rpc_address = fake_socket;
        validator_config.metrics_address = fake_socket;
        validator_config.admin_interface_port = 8888;
        let metrics_addr: Multiaddr = "/ip4/127.0.0.1/tcp/1234".parse().unwrap();
        if let Some(consensus_config) = validator_config.consensus_config.as_mut() {
            consensus_config.consensus_address = Multiaddr::empty();
            consensus_config.consensus_db_path = PathBuf::from("/tmp/foo/");
            consensus_config
                .narwhal_config
                .consensus_api_grpc
                .socket_addr = Multiaddr::empty();
            consensus_config
                .narwhal_config
                .prometheus_metrics
                .socket_addr = metrics_addr
        }
    }
    assert_yaml_snapshot!(network_config, {
        ".genesis" => "[fake genesis]",
        ".validator_configs[].genesis.genesis" => "[fake genesis]",
    });
}
