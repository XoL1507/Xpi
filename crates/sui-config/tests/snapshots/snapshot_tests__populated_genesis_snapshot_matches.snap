---
source: crates/sui-config/tests/snapshot_tests.rs
expression: genesis.validator_set()
---
- name: "0"
  public-key: IP26ybELdYe7p7W8FjvOaeeW1x5O1EwQ/LRIhon3oUQ=
  stake: 1
  delegation: 0
  network-address: ""
  narwhal-primary-to-primary: ""
  narwhal-worker-to-primary: ""
  narwhal-primary-to-worker: ""
  narwhal-worker-to-worker: ""
  narwhal-consensus-address: ""

